pinot-core/src/main/java/org/apache/pinot/core/operator/docidsets/AndDocIdSet.java

```diff
@@ -75,13 +77,22 @@ public BlockDocIdIterator iterator() {
     List<BitmapBasedDocIdIterator> bitmapBasedDocIdIterators = new ArrayList<>();
     List<ScanBasedDocIdIterator> scanBasedDocIdIterators = new ArrayList<>();
     List<BlockDocIdIterator> remainingDocIdIterators = new ArrayList<>();
-    for (int i = 0; i < numDocIdSets; i++) {
-      BlockDocIdIterator docIdIterator = _docIdSets.get(i).iterator();
+
+    Iterator<BlockDocIdSet> iterator = _docIdSets.iterator();
+    for (int i = 0; iterator.hasNext(); i++) {
+      BlockDocIdSet blockDocIdSet = iterator.next();
+      BlockDocIdIterator docIdIterator = blockDocIdSet.iterator();
       allDocIdIterators[i] = docIdIterator;
       if (docIdIterator instanceof SortedDocIdIterator) {
         sortedDocIdIterators.add((SortedDocIdIterator) docIdIterator);
+        // do not keep holding on to the _docIdRanges since they will occupy heap space during the query execution
+        _numEntriesScannedInFilter += blockDocIdSet.getNumEntriesScannedInFilter();
```

Comment:
I don't think we need to collect `numEntriesScannedInFilter` here. When creating the iterator, no scan has happened yet. Same for other places

Reply:
Yeah should be cleaner w/o this, I was actually thinking the same... The problem was `H3IndexQueriesTest` would fail because it produces a `BitmapBasedDocIdIterator` with non-zero `numEntriesScannedInFilter` calculated before hand; therefore, I added this. I think conceptually for `SortedDocIdIterator` and `BitmapBasedDocIdIterator`, when they are generated by a special index then the `numEntriesScannedInFilter` should be already available at the creation of this doc id set; so we can aggregate them here. WDYT @Jackie-Jiang. I don't want to break the stats for these indices.

Reply:
Sounds good. Please add some comments explaining this

Reply:
Added